### PR TITLE
Develop

### PR DESCRIPTION
**Pull request by vladyslav-ko**

### CODE REVIEW

File: src/main/java/com/reportportal/qa/testbase/CucumberHooks.java [hunk 1]
```diff
@@ -3,24 +3,47 @@
...
-    }
+public class CucumberHooks {
 
     @Before("@WebTests")
     public void openUrl() {
         ConfigReader.readProperties(Constants.CONFIGURATION_FILEPATH);
         Selenide.open(ConfigReader.getProperty("url"));
     }
 
+    @After("@WebTests")
```
Comment (dkatsechau) [inline src/main/java/com/reportportal/qa/testbase/CucumberHooks.java:21]: What is a semantic difference between @WebTests and @selenium?
Please, specify more logical and obvious tagging approach.

File: src/main/java/com/reportportal/qa/testbase/CucumberHooks.java [hunk 2]
```diff
@@ -3,24 +3,47 @@
...
+        }
+        Selenide.closeWebDriver();
+    }
+
     @Before("@ApiTests")
     public void setVariables() {
         ConfigReader.readProperties(Constants.CONFIGURATION_FILEPATH);
     }
+
+    @Before("@selenium")
```
Comment (dkatsechau) [inline src/main/java/com/reportportal/qa/testbase/CucumberHooks.java:35]: All your existing tags are in UpperCamelCase. Why not here?

File: src/main/java/com/reportportal/qa/testbase/CucumberHooks.java [hunk 3]
```diff
@@ -3,24 +3,47 @@
...
+    public void openSeleniumLink() {
+        ConfigReader.readProperties(Constants.CONFIGURATION_FILEPATH);
+        getDriver().get(ConfigReader.getProperty("url"));
+    }
+
+    @After("@selenium")
+    public void closeSeleniumBrowser(Scenario scenario) {
```
Comment (dkatsechau) [inline src/main/java/com/reportportal/qa/testbase/CucumberHooks.java:42]: A lot of code duplication. You can arrange your code differently - e.g. implementing interface to manage different browser framework libraries. Or just any conditional/switch structure.

File: src/main/java/com/reportportal/qa/testbase/CucumberHooks.java [hunk 4]
```diff
@@ -3,24 +3,47 @@
...
     public void openUrl() {
         ConfigReader.readProperties(Constants.CONFIGURATION_FILEPATH);
         Selenide.open(ConfigReader.getProperty("url"));
     }
 
+    @After("@WebTests")
+    public void closeBrowser(Scenario scenario) {
+        if (scenario.isFailed()) {
+            String screenshotName = scenario.getName().replaceAll(" ", "_");
+            ScreenshotUtils.takeScreenshot(screenshotName);
```
Comment (dkatsechau) [inline src/main/java/com/reportportal/qa/testbase/CucumberHooks.java:25]: Screenshot is taken on the **After** stage, not on the verification failure.
Not clear, why we need to take screenshot in the **closeBrowser** method.

File: src/main/java/com/reportportal/qa/definitions/LoginPageSeleniumDefinitions.java [hunk 5]
```diff
@@ -0,0 +1,17 @@
+package com.reportportal.qa.definitions;
+
+import com.reportportal.qa.steps.LoginPageSeleniumSteps;
+import io.cucumber.java.en.Given;
+import org.assertj.core.api.Assertions;
+
+public class LoginPageSeleniumDefinitions {
+
+    private final LoginPageSeleniumSteps loginPageSteps = new LoginPageSeleniumSteps();
+
```
Comment (dkatsechau) [inline src/main/java/com/reportportal/qa/definitions/LoginPageSeleniumDefinitions.java:10]: Redundant empty line.

File: src/main/java/com/reportportal/qa/steps/WidgetsPageSeleniumSteps.java [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+
+import static com.reportportal.qa.utils.LaunchDataKey.ELEMENT_HEIGHT;
+import static com.reportportal.qa.utils.LaunchDataKey.ELEMENT_POSITION_X;
+import static com.reportportal.qa.utils.LaunchDataKey.ELEMENT_POSITION_Y;
+import static com.reportportal.qa.utils.LaunchDataKey.ELEMENT_WIDTH;
+
+import com.reportportal.qa.pages.WidgetsSeleniumPage;
+import com.reportportal.qa.utils.DataContext;
+
+public class WidgetsPageSeleniumSteps {
```
Comment (dkatsechau) [inline src/main/java/com/reportportal/qa/steps/WidgetsPageSeleniumSteps.java:11]: Any logging here?

File: src/main/java/com/reportportal/qa/steps/JavaScriptAndActionsSteps.java [hunk 7]
```diff
@@ -0,0 +1,40 @@
+package com.reportportal.qa.steps;
+
+import static com.reportportal.qa.testbase.DriverBuilder.getDriver;
+
+import org.openqa.selenium.JavascriptExecutor;
+import org.openqa.selenium.WebDriver;
+import org.openqa.selenium.WebElement;
+import org.openqa.selenium.interactions.Actions;
+
+public class JavaScriptAndActionsSteps {
```
Comment (dkatsechau) [inline src/main/java/com/reportportal/qa/steps/JavaScriptAndActionsSteps.java:10]: Any logging here?

Comment (dkatsechau) [inline src/main/java/com/reportportal/qa/steps/JavaScriptAndActionsSteps.java:10]: Any logging here?

File: src/main/java/com/reportportal/qa/steps/WidgetsPageSeleniumSteps.java (hunk 6 above)
Comment (dkatsechau) [inline src/main/java/com/reportportal/qa/steps/WidgetsPageSeleniumSteps.java:11]: Code duplication again - you have different approach for steps/definitions for Selenium and Selenide.

**Review by dkatsechau — CHANGES_REQUESTED** [review]

File: src/test/java/SeleniumTestRunner.java [hunk 8]
```diff
@@ -0,0 +1,19 @@
+import io.cucumber.testng.AbstractTestNGCucumberTests;
+import io.cucumber.testng.CucumberOptions;
+import org.testng.annotations.DataProvider;
+
+@CucumberOptions(
+    features = "src/test/java/features",
+    glue = "com/reportportal/qa",
+    tags = "@selenium",
```
[the remark this answers is not in the document]
Comment (vladyslav-ko) [reply src/test/java/SeleniumTestRunner.java]: Now I have one TestRunner

File: src/main/java/com/reportportal/qa/testbase/CucumberHooks.java (hunk 1 above)
Comment (vladyslav-ko) [reply src/main/java/com/reportportal/qa/testbase/CucumberHooks.java]: I changed these tags, "@webtests" - "@Selenide", "selenium" - "@Selenium"

File: src/main/java/com/reportportal/qa/testbase/CucumberHooks.java (hunk 2 above)
Comment (vladyslav-ko) [reply src/main/java/com/reportportal/qa/testbase/CucumberHooks.java]: I changed these tags, "@webtests" - "@Selenide", "selenium" - "@Selenium"

File: src/main/java/com/reportportal/qa/testbase/CucumberHooks.java (hunk 3 above)
Comment (vladyslav-ko) [reply src/main/java/com/reportportal/qa/testbase/CucumberHooks.java:42]: I will do this in the future, don't have time for this now

File: src/main/java/com/reportportal/qa/testbase/CucumberHooks.java (hunk 4 above)
Comment (vladyslav-ko) [reply src/main/java/com/reportportal/qa/testbase/CucumberHooks.java:25]: If we have verification failure, then After stage will run, and that's why before close/quite browser we make a screenshot

File: src/main/java/com/reportportal/qa/definitions/LoginPageSeleniumDefinitions.java (hunk 5 above)
Comment (vladyslav-ko) [reply src/main/java/com/reportportal/qa/definitions/LoginPageSeleniumDefinitions.java:10]: I deleted it

File: src/main/java/com/reportportal/qa/steps/WidgetsPageSeleniumSteps.java (hunk 6 above)
Comment (vladyslav-ko) [reply src/main/java/com/reportportal/qa/steps/WidgetsPageSeleniumSteps.java:11]: I will do this in the future, don't have time for this now

File: src/main/java/com/reportportal/qa/steps/JavaScriptAndActionsSteps.java (hunk 7 above)
Comment (vladyslav-ko) [reply src/main/java/com/reportportal/qa/steps/JavaScriptAndActionsSteps.java:10]: I will do this in the future, don't have time for this now

File: src/main/java/com/reportportal/qa/steps/WidgetsPageSeleniumSteps.java (hunk 6 above)
Comment (dkatsechau) [reply src/main/java/com/reportportal/qa/steps/WidgetsPageSeleniumSteps.java:11]: One of the course objectives - to learn, understand and use code design principles, not only to use new tools.

File: src/main/java/com/reportportal/qa/testbase/CucumberHooks.java (hunk 4 above)
Comment (dkatsechau) [reply src/main/java/com/reportportal/qa/testbase/CucumberHooks.java:25]: It is still very conditional and interdependent implementation. Independent solution is more suitable here.

File: src/main/java/com/reportportal/qa/testbase/CucumberHooks.java (hunk 1 above)
Comment (dkatsechau) [reply src/main/java/com/reportportal/qa/testbase/CucumberHooks.java]: Are selenide tests not "web"?

**Comment by dkatsechau** [comment]
I am not able to properly review PR with such an entangled commits structure. Please, use git best practices, squash commits and give them meaningful names.

**Review by dkatsechau — APPROVED** [review]

**Pull request merged**